python/ray/autoscaler/v2/instance_manager/instance_manager.py

```diff
@@ -0,0 +1,296 @@
+from collections import defaultdict
+import copy
+import logging
+import time
+import uuid
+from abc import ABCMeta, abstractmethod
+from typing import Any, Dict, List, Optional
+
+from ray.autoscaler.v2.instance_manager.instance_storage import (
+    InstanceStorage,
+    InstanceUpdatedSubscriber,
+)
+from ray.core.generated.instance_manager_pb2 import (
+    Status,
+    GetInstanceManagerStateReply,
+    Instance,
+    InstanceManagerState,
+    UpdateInstanceManagerStateReply,
+    UpdateInstanceManagerStateRequest,
+    LaunchRequest,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class InstanceManager(metaclass=ABCMeta):
+    """
+    See `InstanceManagerService` in instance_manager.proto
+    """
+
+    @abstractmethod
+    def update_instance_manager_state(
+        self, request: UpdateInstanceManagerStateRequest
+    ) -> UpdateInstanceManagerStateReply:
+        pass
+
+    @abstractmethod
+    def get_instance_manager_state(self) -> GetInstanceManagerStateReply:
+        pass
+
+
+class SimpleInstanceManager(InstanceManager):
+    """Simple instance manager that manages instances in memory.
+
+    This handles the following updates to an instance:
+        1. when creating new instances
+            An instance is created from an autoscaler's launch request.
+            This initializes an instance object with:
+            status = (Instance.UNKNOWN, Instance.RAY_STATUS_UNKNOWN)
+        2. when ray is stopping
+            This happens when the autoscaler is terminating the ray
+            process on the instance, e.g. idle termination.
+            status = (Instance.ALLOCATED, Instance.RAY_STOPPING)
+        3. when ray is already stopped.
+            Only the ray cluster has the true status of the ray process
+            on an instance, so autoscaler will update an instance's ray
+            to be stopped.
+            status = (Instance.ALLOCATED, Instance.RAY_STOPPED)
+
+        The instance manager should not update the instance status directly,
```

Comment:
is this specific to this implementation? Or should we mention it in the top-level interface?

Reply:
Maybe not specific to this implementation. So yeah, I can move it somewher eelse